backends/ebpf/ebpfenv.py

```diff
@@ -0,0 +1,162 @@
+        """ Append a command to an open process. """
+        return self.ns_proc_write(proc, " && " + cmd)
+
+    def ns_proc_close(self, proc):
+        """ Close and actually run the process in the namespace. Returns the
+            exit code. """
+        errmsg = ("Failed to execute the command"
+                  " sequence in namespace %s" % self.ns_name)
+        return run_process(self.verbose, proc, TIMEOUT, self.outputs, errmsg)
+
+    def _configure_bridge(self, br_name):
+        """ Set the bridge active and load the qdisc. We also disable IPv6 to
+            avoid ICMPv6 spam. """
+        cmd = "ip link set dev %s up && " % br_name
+        # Prevent the broadcasting of ipv6 link discovery messages
+        cmd += "sysctl -w net.ipv6.conf.all.disable_ipv6=1 && "
+        cmd += "sysctl -w net.ipv6.conf.default.disable_ipv6=1 && "
+        # Add the qdisc. MUST be clsact layer.
+        cmd += "tc qdisc add dev %s clsact" % br_name
+        return self.ns_exec(cmd)
+
+    def create_bridge(self):
+        """ Create the central bridge of the environment and configure it. """
+        result = self.ns_exec("ip link add %s type bridge" % self.br_name)
+        if result != SUCCESS:
+            return result
+        return self._configure_bridge(self.br_name)
+
+    def _configure_bridge_port(self, port_name):
+        """ Set a bridge port active and load the qdisc. """
+        cmd = "ip link set dev %s up && " % port_name
+        cmd += "tc qdisc add dev %s clsact" % port_name
+        return self.ns_exec(cmd)
+
+    def attach_interfaces(self, num_ifaces):
+        """ Attach and initialize n interfaces to the central bridge of the
+            namespace. """
+        for index in (range(num_ifaces)):
+            edge_tap = "%s" % index
+            result = self.ns_exec("ip link add link %s name %s"
+                                  " type macvtap mode vepa"
+                                  % (self.br_name, edge_tap))
+            if result != SUCCESS:
+                return result
+            result = self._configure_bridge_port(edge_tap)
+            if result != SUCCESS:
+                return result
+            # add interface to the list of existing bridge ports
+            self.br_ports.append(edge_tap)
+        return SUCCESS
+
+    def create_virtual_env(self, num_ifaces):
+        """ Create the namespace, the bridge, and attach interfaces all at
+            once. """
+        result = self.ns_init()
+        if result != SUCCESS:
+            return result
+        result = self.create_bridge()
+        if result != SUCCESS:
+            return result
+        return self.attach_interfaces(num_ifaces)
+
+
+def main(argv):
+    """ main """
+    # This is a simple test script which creates/deletes a virtual environment
+    outputs = {}
+    outputs["stdout"] = sys.stdout
+    outputs["stderr"] = sys.stderr
+    bridge = Bridge("12345", outputs, True)
+    bridge.create_virtual_env(5)
```

Comment:
you should probably check return codes here

Reply:
The comment says ‘test script’. If it is a test script it should fail when things are not right.